lnrpc/invoicesrpc/addinvoice_test.go

```diff
@@ -0,0 +1,575 @@
+			RemoteBalance:  100,
+			ShortChannelID: private3ShortID,
+		}
+		private3ChannelID = lnwire.NewChanIDFromOutPoint(
+			&privateChannel3.FundingOutpoint,
+		)
+
+		// Create a edge policy for private channel 1.
+		privateChan3Policy = &channeldb.ChannelEdgePolicy{
+			FeeBaseMSat:               30,
+			FeeProportionalMillionths: 300,
+			TimeLockDelta:             3000,
+		}
+
+		// Create a hop hint based on privateChan2Policy.
+		privateChannel3Hint = zpay32.HopHint{
+			NodeID:      privateChannel3.RemotePubkey,
+			ChannelID:   private3ShortID.ToUint64(),
+			FeeBaseMSat: uint32(privateChan3Policy.FeeBaseMSat),
+			FeeProportionalMillionths: uint32(
+				privateChan3Policy.FeeProportionalMillionths,
+			),
+			CLTVExpiryDelta: privateChan3Policy.TimeLockDelta,
+		}
+	)
+
+	// We can't copy in the above var decls, so we copy in our pubkey here.
+	var peer [33]byte
+	copy(peer[:], compressed)
+
+	var (
+		// We pick our policy based on which node (1 or 2) the remote
+		// peer is. Here we create two different sets of edge
+		// information. One where our peer is node 1, the other where
+		// our peer is edge 2. This ensures that we always pick the
+		// right edge policy for our hint.
+		infoNode1 = &channeldb.ChannelEdgeInfo{
+			NodeKey1Bytes: peer,
+		}
+
+		infoNode2 = &channeldb.ChannelEdgeInfo{
+			NodeKey1Bytes: [33]byte{9, 9, 9},
+			NodeKey2Bytes: peer,
+		}
+
+		// setMockChannelUsed preps our mock for the case where we
+		// want our private channel to be used for a hop hint.
+		setMockChannelUsed = func(h *hopHintsConfigMock,
+			channelID lnwire.ChannelID,
+			shortID lnwire.ShortChannelID,
+			policy *channeldb.ChannelEdgePolicy) {
+
+			// Return channel is active so we'll consider it for
+			// hints.
+			h.Mock.On(
+				"IsChannelActive", channelID,
+			).Once().Return(true)
+
+			// Return public node = true so that we'll consider
+			// this node for our hop hints.
+			h.Mock.On(
+				"IsPublicNode", peer,
+			).Once().Return(true, nil)
+
+			// When it gets time to find an edge policy for this
+			// node, fail it. We won't use it as a hop hint.
+			h.Mock.On(
+				"FetchChannelEdgesByID",
+				shortID.ToUint64(),
+			).Once().Return(
+				infoNode1, policy, otherChanPolicy, nil,
+			)
+		}
+	)
+
+	tests := []struct {
+		name      string
+		setupMock func(*hopHintsConfigMock)
+		amount    lnwire.MilliSatoshi
+		channels  []*HopHintInfo
+		numHints  int
+
+		// expectedHints is the set of hop hints that we expect. We
+		// initialize this slice with our max hop hints length, so this
+		// value won't be nil even if its empty.
+		expectedHints [][]zpay32.HopHint
+	}{
+		{
+			// We don't need hop hints for public channels.
+			name: "channel is public",
+			// When a channel is public, we exit before we make any
+			// calls.
+			setupMock: func(h *hopHintsConfigMock) {
+			},
+			amount: 100,
+			channels: []*HopHintInfo{
+				publicChannel,
+			},
+			numHints:      2,
+			expectedHints: make([][]zpay32.HopHint, 0, 2),
```

Comment:
Does `require.Equal` check the capacity? if so we check the length of the `expectedHints` during the test and skip the defining empty slices with the right capacity here.

```
                        hints := SelectHopHints(
				test.amount, cfg, test.channels, test.numHints,
			)
                        require.Equal(t, len(test.expectedHints), len(hints))
                        if len(test.expectedHints) > 0 {
                            require.Equal(t, test.expectedHints, hints)
                        }
```

Reply:
> Does require.Equal check the capacity

It does, because we have `[][]route.Hint{nil}` rather than a nil entry. 

> skip the defining empty slices with the right capacity here

heh yeah, I was losing the fight against testify at this point :') updated to check empty expected slice